presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java

```diff
@@ -81,10 +81,6 @@ public Block getRegion(int position, int length)
             throw new IndexOutOfBoundsException(format("Invalid position range [%s, %s) in block with %s positions", position, position + length, positionCount));
         }
 
-        if (position == 0 && length == positionCount) {
-            return this;
```

Comment:
Isn't that it was correct behavior for an `ArrayBlock`, but unexpected for an `ArrayBlockBuilder`?
`ArrayBlockBuilder` could simply override this method. You can also move this method to  `ArrayBlock` as unsuitable for one of the subclasses.

Reply:
@findepi : This could be another solution. Since `Slice.slice(0, lengh)` will return `this`. We just want to make sure `BlockBuilder` doesn't return `this`.

```java
    /**
     * Returns a slice of this buffer's sub-region. Modifying the content of
     * the returned buffer or this buffer affects each other's content.
     */
    public Slice slice(int index, int length)
    {
        if ((index == 0) && (length == length())) {
            return this;
        }
        checkIndexLength(index, length);
        if (length == 0) {
            return Slices.EMPTY_SLICE;
        }

        if (reference == COMPACT) {
            return new Slice(base, address + index, length, retainedSize, NOT_COMPACT);
        }
        return new Slice(base, address + index, length, retainedSize, reference);
    }
``` 

thoughts @haozhun ?